text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+
+```rust
+extern "crabi" fn func(a: u32, t: cr#(u64, u16), b: i8) -> cr#(u8, u32);
+```
+
+is equivalent to:
+
+```c
+struct func_t_arg {
+    uint64_t f1;
+    uint16_t f2;
+};
+struct func_ret {
+    uint8_t f1;
+    uint32_t f2;
+};
+extern struct func_ret func(uint32_t a, struct func_t_arg t, int8_t b);
+```
+
+Note that crABI tuples are not the same type as Rust tuples, as Rust tuples do
+not guarantee a stable layout.
+
+## `&[T]` or `&mut [T]` - borrowed slice
+
+crABI supports slices (references to arrays) as long as crABI supports the
+element type.
+
+Slices translate to a by-value struct containing two fields: a pointer to the
+element type, and a `size_t` number of elements, in that order.
+
+For instance:
+
+```rust
+extern "crabi" fn func(buf: &mut [u8]);
+```
+
+is equivalent to:
+
+```c
+struct u8_slice {
+    uint8_t *data;
+    size_t len;
+};
+extern void func(struct u8_slice buf);
+```
+
+## `&str` - borrowed counted UTF-8 string
+
+crABI supports `&str`, a borrowed counted UTF-8 string. This type must always
+contain valid UTF-8. (Use a different type for non-UTF-8 strings.) These are
+not typically NUL-terminated.
+
+crABI handles this type equivalently to a byte slice `&[u8]`.
+
+## `Box<[T]>` - boxed slice
+
+crABI supports passing ownership of a slice (an array with size determined at
+runtime). To represent this via the C ABI, crABI passes this equivalently to a
+C struct containing an element pointer and a `size_t` number of elements, in
+that order.
+
+For instance:
+
+```rust
+extern "crabi" fn func(data: Box<[u32]>) -> Box<[f64]>;
+```
+
+is equivalent to:
+
+```c
+struct u32_boxed_slice {
+    uint32_t *data;
+    size_t len;
+};
+struct f64_boxed_slice {
+    double *data;
+    size_t len;
+};
+extern struct f64_boxed_slice func(struct u32_boxed_slice data);
+```
+
+## `Box<str>` - boxed counted UTF-8 string
+
+crABI supports passing an owned string. To represent this via the C ABI, crABI
+passes this equivalently to a boxed slice of type `Box<[u8]>`.
+
+## `[T; N]` - fixed-size array by value
+
+crABI supports passing a fixed-size array by value (as opposed to by
+reference). To represent this via the C ABI, crABI treats this equivalently to
+a C array of the same type passed by value.
+
+Note that this means C code can use an array directly in a context where it
+will be interpreted by-value (such as in a struct field), but needs to use a
+structure with a single array field in contexts where it would otherwise be
+interpreted as a pointer (such as in a function argument or return value).
+
+For instance:
+
+```rust
+extern "crabi" fn func(rgb: [u16; 3])
+```
+
+is equivalent to:
+
+```c
+struct func_rgb_arg {
+    uint16_t array[3];
+};
+extern void func(struct func_rgb_arg rgb);
+```
+
+Note that crABI does *not* pass the length, since it's a compile-time constant;
+the recipient must also know the correct size. (Use one of the slice-based
+types for a type with a runtime-determined length.)
```

Comment:
I seem to remember issues about zero-sized types in C; what's expected behaviour if `N` is `0`? I'd expect it to act like the other zero-sized types, i.e. the function parameter in it's entirety would just not be there in the C ABI, but that's not entirely clear from the example.

Reply:
I think the `[T; 0]` case might depend on whether the array is passed by reference (address and length) or value (just address). It seems like passing by value is ambiguous here.

Also, what's the expected behaviour for arrays containing zero-sized types?
Is it different for `[(), N]` and `[(), 0]`?
Is it different passing by value and by reference?

Reply:
For the rust abi all ZST's are ignored. Mosy C abi's follow this as well, but on a couple of archs ZST's still consume a register: https://github.com/rust-lang/rust/blob/6ef7d16be0fb9d6ecf300c27990f4bff49d22d46/compiler/rustc_ty_utils/src/abi.rs#L419-L421